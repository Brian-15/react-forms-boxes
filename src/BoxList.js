import { React, useState } from 'react';
import { v4 as uuid } from 'uuid';
import Box from './Box.js';
import NewBoxForm from './NewBoxForm.js';
import './BoxList.css';

function genBoxData(backgroundColor, width, height, id) {
    return { backgroundColor, width, height, id };
}

const defaultData = [
    genBoxData('blue', '500px', '200px', uuid()),
    genBoxData('green', '50px', '60px', uuid()),
    genBoxData('red', '50px', '60px', uuid()),
]

const BoxList = ({ initialData=defaultData }) => {
    const [ boxData, setBoxData ] = useState(initialData);

    const addBox = data => {
        setBoxData(boxData => [...boxData, {...data, id: uuid()}]);
    };

    const removeBox = evt => {
        const id = evt.target.parentElement.id;
        setBoxData(boxData => boxData.filter(box => box.id !== id));
    };

    return (
        <div className='BoxList'>
            <NewBoxForm addBox={addBox} />
            {boxData.map(({id, backgroundColor, width, height}) => (
                <Box
                    id={id}
                    backgroundColor={backgroundColor}
                    width={width}
                    height={height}
                    key={id}
                    removeBox={removeBox}
                />
            ))}
        </div>
    );
};

export default BoxList;