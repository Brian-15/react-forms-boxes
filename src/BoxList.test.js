import { render, fireEvent } from '@testing-library/react';
import BoxList from './BoxList.js';

/** mock uuid with an incrementing id starting from 0
 * from: https://medium.com/@tommybernaciak/a-missing-jest-mock-feature-b89b74d7a09c
 */
jest.mock('uuid', () => ({
    v4: jest.fn().mockImplementation(
      (function(n) {
        return function() {
          return ++n;
        };
      })(0)
    )
}));

test('smoke test: renders BoxList', () => {
    render(<BoxList />);
});

test('matches snapshot', () => {
    const { asFragment } = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
});

test('should add new box', () => {
    const {
        queryAllByTestId,
        queryByTestId,
        queryByText
    } = render(<BoxList />);

    const btn = queryByText('Add Box');
    const color = queryByTestId('bg-color');
    const width = queryByTestId('width');
    const height = queryByTestId('height');

    expect(queryAllByTestId('box').length).toBe(3);

    fireEvent.change(color, { target: { value: 'pink' } });
    fireEvent.change(width, { target: { value: '50px' } });
    fireEvent.change(height, { target: { value: '50px' } });
    fireEvent.click(btn);

    const newBoxes = queryAllByTestId('box');
    expect(newBoxes.length).toBe(4);
    expect(newBoxes[3]).toHaveStyle('background-color: pink;');
});