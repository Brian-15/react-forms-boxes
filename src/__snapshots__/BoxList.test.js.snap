// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <form
      class="NewBoxForm"
      data-testid="form"
    >
      <label
        for="backgroundColor"
      >
        Color:
        <input
          data-testid="bg-color"
          name="backgroundColor"
          value=""
        />
      </label>
      <label
        for="width"
      >
        Width:
        <input
          data-testid="width"
          name="width"
          value=""
        />
      </label>
      <label
        for="height"
      >
        Height:
        <input
          data-testid="height"
          name="height"
          value=""
        />
      </label>
      <button>
        Add Box
      </button>
    </form>
    <div
      class="Box"
      id="1"
    >
      <div
        class="Box-body"
        data-testid="box"
        style="background-color: blue; width: 500px; height: 200px;"
      />
      <button
        class="Box-remove-btn"
      >
        X
      </button>
    </div>
    <div
      class="Box"
      id="2"
    >
      <div
        class="Box-body"
        data-testid="box"
        style="background-color: green; width: 50px; height: 60px;"
      />
      <button
        class="Box-remove-btn"
      >
        X
      </button>
    </div>
    <div
      class="Box"
      id="3"
    >
      <div
        class="Box-body"
        data-testid="box"
        style="background-color: red; width: 50px; height: 60px;"
      />
      <button
        class="Box-remove-btn"
      >
        X
      </button>
    </div>
  </div>
</DocumentFragment>
`;
